framework-t/src/android/net/nsd/NsdManager.java

```diff
@@ -725,7 +726,9 @@ public NsdManager(Context context, INsdManager service) {
         // Instead of launching separate threads to handle tasks from the various instances.
         mHandler = new ServiceHandler(ConnectivityThread.getInstanceLooper());
 
-        if (android.content.pm.SpecialRuntimePermAppUtils.isInternetCompatEnabled()) {
+        int uid = android.os.Process.myUid();
+        if (android.content.pm.SpecialRuntimePermAppUtils.isInternetCompatEnabled() ||
+                ConnectivityUtil.isRegularAppWithLockdownVpnEnabled(mContext, uid)) {
```

Comment:
isRegularAppWithLockdownVpnEnabled() makes several IPC calls to system_server. It's inefficient to perform this check inside an app given that system_server checks the same thing in NsdService.

Instead, NsdService can be made to return null when isRegularAppWithLockdownVpnEnabled() is true. NsdManager can then detect that and fallback to the stubbed out implementation of INsdServiceConnector.